packages/core/src/codewhisperer/tracker/userWrittenCodeTracker.ts

```diff
@@ -0,0 +1,196 @@
+        return this._userWrittenNewCodeLineCount.get(language) || 0
+    }
+
+    public reset() {
+        this._userWrittenNewCodeLineCount = new Map<CodewhispererLanguage, number>()
+        this._userWrittenNewCodeCharacterCount = new Map<CodewhispererLanguage, number>()
+        this._qUsageCount = 0
+        this._qIsMakingEdits = false
+        this._lastQInvocationTime = 0
+        if (this._timer !== undefined) {
+            clearTimeout(this._timer)
+            this._timer = undefined
+        }
+    }
+
+    public emitCodeContributions() {
+        const selectedCustomization = getSelectedCustomization()
+
+        for (const [language, charCount] of this._userWrittenNewCodeCharacterCount) {
+            const lineCount = this._userWrittenNewCodeLineCount.get(language) || 0
+            if (charCount > 0) {
+                client
+                    .sendTelemetryEvent({
+                        telemetryEvent: {
+                            codeCoverageEvent: {
+                                customizationArn:
+                                    selectedCustomization.arn === '' ? undefined : selectedCustomization.arn,
+                                programmingLanguage: {
+                                    languageName: runtimeLanguageContext.toRuntimeLanguage(language),
+                                },
+                                acceptedCharacterCount: 0,
+                                totalCharacterCount: 0,
+                                timestamp: new Date(Date.now()),
+                                userWrittenCodeCharacterCount: charCount,
+                                userWrittenCodeLineCount: lineCount,
+                            },
+                        },
+                    })
+                    .then()
+                    .catch((error) => {
+                        let requestId: string | undefined
+                        if (isAwsError(error)) {
+                            requestId = error.requestId
+                        }
+                        getLogger().debug(
+                            `Failed to sendTelemetryEvent, requestId: ${requestId ?? ''}, message: ${error.message}`
+                        )
+                    })
+            }
+        }
+    }
+
+    private tryStartTimer() {
+        if (this._timer !== undefined) {
+            return
+        }
+        if (!this.isActive()) {
+            getLogger().debug(`Skip emiting code contribution metric. Telemetry disabled or not logged in. `)
+            this.reset()
+            return
+        }
+        const startTime = performance.now()
+        this._timer = setTimeout(() => {
+            try {
+                const currentTime = performance.now()
+                const delay: number = UserWrittenCodeTracker.defaultCheckPeriodMillis
+                const diffTime: number = startTime + delay
+                if (diffTime <= currentTime) {
```

Comment:
what's this for? is this for nodejs@setTimeout 
```
The callback will likely not be invoked in precisely delay milliseconds. Node.js makes no guarantees about the exact timing of when callbacks will fire, nor of their ordering. The callback will be called as close as possible to the time specified.
```
https://nodejs.org/api/timers.html#settimeoutcallback-delay-args

However by doing this, doesn't it mean some metrics will not be sent out

Reply:
This implementation of per 5 min time window is exactly the same as previous code percentage tracker.